File: api/tts.js
// Vercel Edge TTS API代理服务 - 简化版本
// 解决ES模块兼容性问题

import { WebSocket } from 'ws';
import { createHmac, randomUUID } from 'crypto';

// Vercel Serverless Function 主入口
export default async function handler(req, res) {
    // 处理 CORS 预检请求
    if (req.method === 'OPTIONS') {
        res.setHeader('Access-Control-Allow-Origin', '*');
        res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
        return res.status(200).end();
    }
    
    // 设置 CORS 响应头
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method Not Allowed' });
    }

    try {
        const { text, voice = 'zh-CN-XiaochenNeural', rate = 0, pitch = 0, volume = 50 } = req.body;

        if (!text || !text.trim()) {
            return res.status(400).json({ success: false, message: '请提供要转换的文本' });
        }

        console.log('TTS请求:', { textLength: text.length, voice });
        
        // 文本预处理
        const textChunks = preprocessText(text);
        console.log(`文本分为 ${textChunks.length} 个块`);

        // 获取TTS服务端点
        const { endpoint } = await getTTSEndpoint();
        const trustedToken = endpoint.t;

        const allAudioChunks = [];
        let allWordBoundaries = [];
        let totalAudioDurationNs = 0;

        // 处理每个文本块
        for (let i = 0; i < textChunks.length; i++) {
            const chunk = textChunks[i];
            console.log(`处理第 ${i + 1}/${textChunks.length} 个文本块`);
            
            const ssml = generateSSML(chunk, voice, rate, pitch, volume);
            const webSocketUrl = `wss://speech.platform.bing.com/consumer/speech/synthesize/readaloud/edge/v1?TrustedClientToken=${trustedToken}`;
            const clientId = randomUUID().replace(/-/g, '');
            const webSocketSignature = await generateWebSocketSignature(webSocketUrl);

            try {
                const result = await getAudioAndSubtitlesFromWebSocket(webSocketUrl, webSocketSignature, clientId, ssml);
                
                allAudioChunks.push(result.audioBuffer);

                // 校正字幕时间戳
                const correctedBoundaries = result.wordBoundaries.map(wb => ({
                    ...wb,
                    AudioOffset: wb.AudioOffset + totalAudioDurationNs
                }));
                allWordBoundaries = allWordBoundaries.concat(correctedBoundaries);

                // 更新总时长
                if (result.wordBoundaries.length > 0) {
                    const lastWord = result.wordBoundaries[result.wordBoundaries.length - 1];
                    totalAudioDurationNs += lastWord.AudioOffset + lastWord.AudioDuration;
                }
            } catch (chunkError) {
                console.error(`处理第 ${i + 1} 个文本块失败:`, chunkError);
                throw new Error(`处理文本块 ${i + 1} 失败: ${chunkError.message}`);
            }
        }

        // 合并结果
        const finalAudioBuffer = Buffer.concat(allAudioChunks);
        const audioBase64 = finalAudioBuffer.toString('base64');
        const srtContent = generateAccurateSRT(allWordBoundaries);
        const actualDuration = totalAudioDurationNs / 10000000;

        console.log('TTS处理完成:', { audioSize: finalAudioBuffer.length, duration: actualDuration });

        return res.status(200).json({
            success: true,
            audio_base64: audioBase64,
            srt_string: srtContent,
            message: '语音生成成功',
            voice, rate, pitch, volume,
            duration: actualDuration,
            chunks: textChunks.length
        });

    } catch (error) {
        console.error('Vercel TTS处理错误:', error);
        return res.status(500).json({ 
            success: false, 
            message: `语音合成失败: ${error.message}`
        });
    }
}

// ==============================================================================
// 文本预处理函数
// ==============================================================================
function preprocessText(text) {
    // 移除标点符号，保留中文、字母、数字和空格
    const cleanedText = text.replace(/[^\p{L}\p{N}\s]/gu, '');
    
    // 按句子分割
    const sentences = cleanedText.split(/(?<=[。！？.])\s*/).filter(s => s.trim());
    const chunks = [];
    const maxLength = 200;

    sentences.forEach(sentence => {
        if (sentence.length <= maxLength) {
            chunks.push(sentence);
        } else {
            // 超长句子进行硬切分
            for (let i = 0; i < sentence.length; i += maxLength) {
                chunks.push(sentence.substring(i, i + maxLength));
            }
        }
    });

    return chunks.filter(c => c.trim());
}

// ==============================================================================
// WebSocket通信核心逻辑
// ==============================================================================
function getAudioAndSubtitlesFromWebSocket(url, signature, clientId, ssml) {
    return new Promise((resolve, reject) => {
        const headers = {
            'X-ClientTraceId': clientId,
            'X-MT-Signature': signature,
            'Origin': 'chrome-extension://jdiccldimpdaibmpdkjnbmckianbfold',
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36 Edg/91.0.864.41'
        };

        const ws = new WebSocket(url, {
            headers: headers
        });

        let audioChunks = [];
        let wordBoundaries = [];
        let responseSent = false;

        // 设置30秒超时
        const timeout = setTimeout(() => {
            if (!responseSent) {
                responseSent = true;
                ws.close();
                reject(new Error('WebSocket操作超时'));
            }
        }, 30000);

        ws.on('open', () => {
            console.log('WebSocket连接已建立');
            
            // 发送配置消息
            const configMessage = {
                context: {
                    synthesis: {
                        audio: {
                            metadataoptions: {
                                sentenceBoundaryEnabled: 'false',
                                wordBoundaryEnabled: 'true'
                            },
                            outputFormat: 'audio-24khz-48kbitrate-mono-mp3'
                        }
                    }
                }
            };

            const configPayload = `X-Timestamp:${new Date().toISOString()}\r\nPath:speech.config\r\nContent-Type:application/json; charset=utf-8\r\n\r\n${JSON.stringify(configMessage)}`;
            ws.send(configPayload);

            // 发送SSML消息
            const ssmlPayload = `X-Timestamp:${new Date().toISOString()}\r\nPath:ssml\r\nContent-Type:application/ssml+xml\r\n\r\n${ssml}`;
            ws.send(ssmlPayload);
        });

        ws.on('message', (data) => {
            try {
                if (typeof data === 'string') {
                    const lines = data.split('\r\n');
                    const pathLine = lines.find(line => line.startsWith('Path:'));
                    
                    if (pathLine && pathLine.includes('wordBoundary')) {
                        const jsonLine = lines.find(line => line.startsWith('{'));
                        if (jsonLine) {
                            const boundary = JSON.parse(jsonLine);
                            wordBoundaries.push(boundary);
                        }
                    } else if (pathLine && pathLine.includes('turn.end')) {
                        clearTimeout(timeout);
                        if (!responseSent) {
                            responseSent = true;
                            if (audioChunks.length > 0) {
                                resolve({
                                    audioBuffer: Buffer.concat(audioChunks),
                                    wordBoundaries: wordBoundaries
                                });
                            } else {
                                reject(new Error('未收到音频数据'));
                            }
                        }
                        ws.close();
                    }
                } else if (Buffer.isBuffer(data)) {
                    // 处理二进制音频数据
                    const headerLength = data.readUInt16BE(0);
                    const audioData = data.slice(2 + headerLength);
                    if (audioData.length > 0) {
                        audioChunks.push(audioData);
                    }
                }
            } catch (error) {
                console.error('消息处理错误:', error);
            }
        });

        ws.on('error', (error) => {
            clearTimeout(timeout);
            if (!responseSent) {
                responseSent = true;
                reject(new Error(`WebSocket连接失败: ${error.message}`));
            }
        });

        ws.on('close', (code, reason) => {
            clearTimeout(timeout);
            if (!responseSent) {
                responseSent = true;
                if (audioChunks.length === 0) {
                    reject(new Error(`WebSocket意外关闭。代码: ${code}, 原因: ${reason || '未知'}`));
                } else {
                    resolve({
                        audioBuffer: Buffer.concat(audioChunks),
                        wordBoundaries: wordBoundaries
                    });
                }
            }
        });
    });
}

// ==============================================================================
// 辅助函数
// ==============================================================================
async function getTTSEndpoint() {
    const endpointUrl = 'https://dev.microsofttranslator.com/apps/endpoint?api-version=1.0';
    const signature = await generateTTSSignature(endpointUrl);
    const clientId = randomUUID().replace(/-/g, '');
    
    const headers = {
        'Accept-Language': 'zh-Hans',
        'X-ClientTraceId': clientId,
        'X-MT-Signature': signature,
        'User-Agent': 'okhttp/4.5.0',
        'Content-Type': 'application/json; charset=utf-8'
    };

    const response = await fetch(endpointUrl, {
        method: 'POST',
        headers: headers
    });

    if (!response.ok) {
        throw new Error(`获取TTS Endpoint失败，状态码 ${response.status}`);
    }

    return { endpoint: await response.json(), signature };
}

function generateTTSSignature(urlStr) {
    const url = urlStr.split('://')[1];
    const encodedUrl = encodeURIComponent(url);
    const uuidStr = randomUUID().replace(/-/g, '');
    const formattedDate = new Date().toUTCString();
    const bytesToSign = `MSTranslatorAndroidApp${encodedUrl}${formattedDate}${uuidStr}`.toLowerCase();
    
    const key = Buffer.from('oik6PdDdMnOXemTbwvMn9de/h9lFnfBaCWbGMMZqqoSaQaqUOqjVGm5NqsmjcBI1x+sS9ugjB55HEJWRiFXYFw==', 'base64');
    const hmac = createHmac('sha256', key);
    hmac.update(bytesToSign);
    const signatureBase64 = hmac.digest('base64');
    
    return `MSTranslatorAndroidApp::${signatureBase64}::${formattedDate}::${uuidStr}`;
}

function generateWebSocketSignature(webSocketUrlStr) {
    const url = new URL(webSocketUrlStr);
    const pathAndQuery = url.pathname + url.search;
    const encodedPathAndQuery = encodeURIComponent(pathAndQuery);
    const uuidStr = randomUUID().replace(/-/g, '');
    const formattedDate = new Date().toUTCString();
    const bytesToSign = `MSTranslatorAndroidApp${encodedPathAndQuery}${formattedDate}${uuidStr}`.toLowerCase();
    
    const key = Buffer.from('oik6PdDdMnOXemTbwvMn9de/h9lFnfBaCWbGMMZqqoSaQaqUOqjVGm5NqsmjcBI1x+sS9ugjB55HEJWRiFXYFw==', 'base64');
    const hmac = createHmac('sha256', key);
    hmac.update(bytesToSign);
    const signatureBase64 = hmac.digest('base64');
    
    return `MSTranslatorAndroidApp::${signatureBase64}::${formattedDate}::${uuidStr}`;
}

function generateSSML(text, voiceName, rate, pitch, volume) {
    const escapedText = text
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&apos;');
        
    return `<speak xmlns="http://www.w3.org/2001/10/synthesis" xmlns:mstts="http://www.w3.org/2001/mstts" version="1.0" xml:lang="zh-CN"><voice name="${voiceName}"><mstts:express-as style="general" styledegree="1.0" role="default"><prosody rate="${rate}%" pitch="${pitch}%" volume="${volume}">${escapedText}</prosody></mstts:express-as></voice></speak>`;
}

function generateAccurateSRT(wordBoundaries) {
    let srt = '';
    const groupSize = 12;
    
    for (let i = 0; i < wordBoundaries.length; i += groupSize) {
        const chunk = wordBoundaries.slice(i, i + groupSize);
        if (chunk.length === 0) continue;
        
        const index = Math.floor(i / groupSize) + 1;
        const startTime = chunk[0].AudioOffset / 10000000;
        const lastWord = chunk[chunk.length - 1];
        const endTime = (lastWord.AudioOffset + lastWord.AudioDuration) / 10000000;
        const text = chunk.map(word => word.Text).join('');
        
        srt += `${index}\n${formatTimeSRT(startTime)} --> ${formatTimeSRT(endTime)}\n${text}\n\n`;
    }
    
    return srt;
}

function formatTimeSRT(seconds) {
    const date = new Date(0);
    date.setSeconds(seconds);
    return date.toISOString().substr(11, 12).replace('.', ',');
}

File: api/tts.py
"""
Python TTS代理服务 - 基于edge-tts库
部署到Vercel的无服务器函数

优势:
1. 直接使用成熟的edge-tts Python库
2. 支持精确的字幕时间戳
3. 不需要处理复杂的WebSocket连接
4. 库维护活跃，稳定可靠
"""

import asyncio
import base64
import json
import tempfile
import os
from typing import Dict, Any

try:
    import edge_tts
    from edge_tts import VoicesManager
except ImportError:
    # 如果没有安装，返回错误信息
    edge_tts = None

async def process_tts_request(request_body: Dict[str, Any]) -> Dict[str, Any]:
    """
    Vercel无服务器函数处理器
    """
    try:
        # 检查edge-tts库是否可用
        if edge_tts is None:
            return {
                "statusCode": 500,
                "headers": {
                    "Access-Control-Allow-Origin": "*",
                    "Content-Type": "application/json"
                },
                "body": json.dumps({
                    "success": False,
                    "message": "edge-tts库未安装，请在requirements.txt中添加edge-tts"
                })
            }

        # 解析请求参数
        text = request_body.get('text', '')
        voice = request_body.get('voice', 'zh-CN-XiaochenNeural')
        rate = request_body.get('rate', 0)
        pitch = request_body.get('pitch', 0)
        volume = request_body.get('volume', 50)

        if not text or not text.strip():
            return {
                "statusCode": 400,
                "headers": {
                    "Access-Control-Allow-Origin": "*",
                    "Content-Type": "application/json"
                },
                "body": json.dumps({
                    "success": False,
                    "message": "请提供要转换的文本"
                })
            }

        print(f"处理TTS请求: text_length={len(text)}, voice={voice}")

        # 使用edge-tts生成音频和字幕
        result = await generate_tts_with_subtitles(text, voice, rate, pitch, volume)
        
        return {
            "statusCode": 200,
            "headers": {
                "Access-Control-Allow-Origin": "*",
                "Content-Type": "application/json"
            },
            "body": json.dumps({
                "success": True,
                "audio_base64": result["audio_base64"],
                "srt_string": result["srt_content"],
                "message": "语音生成成功",
                "voice": voice,
                "rate": rate,
                "pitch": pitch,
                "volume": volume,
                "duration": result["duration"],
                "engine": "edge-tts-python"
            })
        }

    except Exception as e:
        print(f"TTS处理错误: {str(e)}")
        return {
            "statusCode": 500,
            "headers": {
                "Access-Control-Allow-Origin": "*",
                "Content-Type": "application/json"
            },
            "body": json.dumps({
                "success": False,
                "message": f"语音生成失败: {str(e)}"
            })
        }


async def generate_tts_with_subtitles(text: str, voice: str, rate: int, pitch: int, volume: int) -> Dict[str, Any]:
    """
    使用edge-tts生成音频和精确字幕
    """
    # 构建SSML参数
    rate_str = f"{rate:+d}%" if rate != 0 else "+0%"
    pitch_str = f"{pitch:+d}Hz" if pitch != 0 else "+0Hz"
    volume_str = f"{volume:+d}%"

    # 创建临时文件
    with tempfile.NamedTemporaryFile(suffix=".mp3", delete=False) as audio_file:
        audio_path = audio_file.name
    
    with tempfile.NamedTemporaryFile(suffix=".vtt", delete=False) as subtitle_file:
        subtitle_path = subtitle_file.name

    try:
        # 使用edge-tts生成音频和字幕
        communicate = edge_tts.Communicate(
            text=text,
            voice=voice,
            rate=rate_str,
            pitch=pitch_str,
            volume=volume_str
        )

        # 同时生成音频和字幕文件
        submaker = edge_tts.SubMaker()
        
        async for chunk in communicate.stream():
            if chunk["type"] == "audio":
                with open(audio_path, "ab") as f:
                    f.write(chunk["data"])
            elif chunk["type"] == "WordBoundary":
                submaker.create_sub((chunk["offset"], chunk["duration"]), chunk["text"])

        # 生成字幕文件
        with open(subtitle_path, "w", encoding="utf-8") as f:
            f.write(submaker.generate_subs())

        # 读取音频文件并转换为base64
        with open(audio_path, "rb") as f:
            audio_data = f.read()
            audio_base64 = base64.b64encode(audio_data).decode('utf-8')

        # 读取字幕文件并转换为SRT格式
        with open(subtitle_path, "r", encoding="utf-8") as f:
            vtt_content = f.read()
            srt_content = convert_vtt_to_srt(vtt_content)

        # 估算音频时长（简单估算）
        duration = len(audio_data) / (128 * 1024 / 8)  # 假设128kbps

        return {
            "audio_base64": audio_base64,
            "srt_content": srt_content,
            "duration": duration
        }

    finally:
        # 清理临时文件
        try:
            os.unlink(audio_path)
            os.unlink(subtitle_path)
        except:
            pass


def convert_vtt_to_srt(vtt_content: str) -> str:
    """
    将VTT字幕转换为SRT格式
    """
    lines = vtt_content.strip().split('\n')
    srt_lines = []
    subtitle_index = 1
    
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        # 跳过VTT头部信息
        if line.startswith('WEBVTT') or line.startswith('NOTE') or not line:
            i += 1
            continue
            
        # 检查是否是时间戳行
        if '-->' in line:
            # 转换时间格式 (00:00:01.000 --> 00:00:02.500)
            time_parts = line.split(' --> ')
            start_time = time_parts[0].replace('.', ',')
            end_time = time_parts[1].replace('.', ',')
            
            # 添加序号和时间戳
            srt_lines.append(str(subtitle_index))
            srt_lines.append(f"{start_time} --> {end_time}")
            
            # 查找并添加字幕文本
            i += 1
            subtitle_text = []
            while i < len(lines) and lines[i].strip() and '-->' not in lines[i]:
                subtitle_text.append(lines[i].strip())
                i += 1
            
            srt_lines.append(' '.join(subtitle_text))
            srt_lines.append('')  # 空行分隔
            subtitle_index += 1
        else:
            i += 1
    
    return '\n'.join(srt_lines)


# Vercel HTTP入口点
from http.server import BaseHTTPRequestHandler
import urllib.parse

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        """处理CORS预检请求"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def do_POST(self):
        """处理POST请求"""
        try:
            # 读取请求体
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            request_body = json.loads(post_data.decode('utf-8'))
            
            # 异步处理TTS请求
            result = asyncio.run(process_tts_request(request_body))
            
            # 发送响应
            self.send_response(result["statusCode"])
            for key, value in result["headers"].items():
                self.send_header(key, value)
            self.end_headers()
            self.wfile.write(result["body"].encode('utf-8'))
            
        except Exception as e:
            self.send_response(500)
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            error_response = json.dumps({
                "success": False,
                "message": f"请求处理失败: {str(e)}"
            })
            self.wfile.write(error_response.encode('utf-8'))
    
    def do_GET(self):
        """处理GET请求 - 返回API信息"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        info = {
            "service": "Python TTS Proxy",
            "status": "running",
            "engine": "edge-tts",
            "method": "POST /api/tts",
            "params": ["text", "voice", "rate", "pitch", "volume"]
        }

        self.wfile.write(json.dumps(info, ensure_ascii=False).encode('utf-8'))
